render/svg/src/lib.rs

```diff
@@ -0,0 +1,868 @@
+        })
+    }
+}
+
+impl fmt::Display for SvgRenderBackend {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.document)
+    }
+}
+
+impl RenderBackend for SvgRenderBackend {
+    fn set_viewport_dimensions(&mut self, width: u32, height: u32) {
+        self.viewport_width = width;
+        self.viewport_height = height;
+        dbg!(width, height);
+        self.document.assign("width", width);
+        self.document.assign("height", height);
+        self.document.assign("viewBox", (0, 0, width, height));
+    }
+
+    fn register_shape(&mut self, shape: DistilledShape) -> ShapeHandle {
+        let handle = ShapeHandle(self.shapes.len());
+
+        let mut bitmaps = HashMap::new();
+        for (id, handle) in &self.id_to_bitmap {
+            let bitmap_data = &self.bitmaps[handle.0];
+            bitmaps.insert(
+                *id,
+                (&bitmap_data.data[..], bitmap_data.width, bitmap_data.height),
+            );
+        }
+
+        let data = swf_shape_to_svg(shape, &bitmaps, "crisp-edges");
+        self.shapes.push(data);
+
+        handle
+    }
+
+    fn replace_shape(&mut self, shape: DistilledShape, handle: ShapeHandle) {
+        let mut bitmaps = HashMap::new();
+        for (id, handle) in &self.id_to_bitmap {
+            let bitmap_data = &self.bitmaps[handle.0];
+            bitmaps.insert(
+                *id,
+                (&bitmap_data.data[..], bitmap_data.width, bitmap_data.height),
+            );
+        }
+
+        let data = swf_shape_to_svg(shape, &bitmaps, "crisp-edges");
+        self.shapes[handle.0] = data;
+    }
+
+    fn register_glyph_shape(&mut self, glyph: &swf::Glyph) -> ShapeHandle {
+        // Per SWF19 p.164, the FontBoundsTable can contain empty bounds for every glyph (reserved).
+        // SWF19 says this is true through SWFv7, but it seems like it might be generally true?
+        // In any case, we have to be sure to calculate the shape bounds ourselves to make a proper
+        // SVG.
+        let bounds = glyph
+            .clone()
+            .bounds
+            .filter(|b| b.x_min != b.x_max || b.y_min != b.y_max)
+            .unwrap_or_else(|| {
+                ruffle_core::shape_utils::calculate_shape_bounds(&glyph.shape_records[..])
+            });
+        let shape = swf::Shape {
+            version: 2,
+            id: 0,
+            shape_bounds: bounds.clone(),
+            edge_bounds: bounds,
+            has_fill_winding_rule: false,
+            has_non_scaling_strokes: false,
+            has_scaling_strokes: true,
+            styles: swf::ShapeStyles {
+                fill_styles: vec![swf::FillStyle::Color(Color {
+                    r: 255,
+                    g: 255,
+                    b: 255,
+                    a: 255,
+                })],
+                line_styles: vec![],
+            },
+            shape: glyph.shape_records.clone(),
+        };
```

Comment:
A recent commit factored this code out into `ruffle_core::shape_utils::swf_glyph_to_shape`; could use that here instead